[APM] Language-specific stacktrace formatting

smith opened the pull request:
Create stackframe heading renderers for languages as described in #49467.

The intention is to make the frames look like they do in their native language stacktrace output.

The [FrameHeading tests](https://github.com/elastic/kibana/pull/75924/files#diff-8d83e358d495d829b26774369f38100c) include examples.

Rename `IStackFrame` to `Stackframe` to match other interfaces.

Fixes #49467.

# Screenshots

## .NET

### Before 

![image](https://user-images.githubusercontent.com/9912/91490854-bba39180-e878-11ea-9880-dae4d1ef34bc.png)

### After

![image](https://user-images.githubusercontent.com/9912/91490879-c65e2680-e878-11ea-894d-1e6786d22c1d.png)


this is without `classname`s. elastic/apm-agent-dotnet#848 is not yet merged, but there are examples with `classname` in the tests.

## Go

### Before 

![image](https://user-images.githubusercontent.com/9912/91491024-09b89500-e879-11ea-92a1-cc8589de26f7.png)

### After

![image](https://user-images.githubusercontent.com/9912/91490982-f1e11100-e878-11ea-8c84-da2a6f305c9d.png)

Only spacing and indentation changed.

## Java

### Before

![image](https://user-images.githubusercontent.com/9912/91491117-37054300-e879-11ea-8095-300fd62bb5be.png)

### After

![image](https://user-images.githubusercontent.com/9912/91491186-4b494000-e879-11ea-9128-e6526a9994fa.png)

## Node

### Before

![image](https://user-images.githubusercontent.com/9912/91491283-6d42c280-e879-11ea-8740-9b9dd18a0996.png)

### After

![image](https://user-images.githubusercontent.com/9912/91491314-7b90de80-e879-11ea-9613-e537dcc4258c.png)

This now includes `classname` if it's sent but we don't have any Opbeans errors that include this.

## Python

Only spacing and indentation changed.

## Ruby

### Before

![image](https://user-images.githubusercontent.com/9912/91491492-b85cd580-e879-11ea-9207-7ed6e551d563.png)

### After

![image](https://user-images.githubusercontent.com/9912/91491418-9c593400-e879-11ea-8148-438588677fcd.png)

## RUM

### Before

![image](https://user-images.githubusercontent.com/9912/91491584-e3dfc000-e879-11ea-8aa4-c3d2149a304d.png)

### After 

![image](https://user-images.githubusercontent.com/9912/91491623-f0fcaf00-e879-11ea-92db-584b4d3f0937.png)

RUM uses the same renderer as Node.

smith committed:
[APM] Language-specific stacktrace formatting

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/70381/)
* Commit: 21ec6b04a5040d35c5637e4c7420c3bebc604bbc
* This comment will update when the build is complete

### Failed CI Steps
* [Check Types](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/70381/execution/node/374/log)

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/70381/","number":"70381","commit":"21ec6b04a5040d35c5637e4c7420c3bebc604bbc"}]}
PIPELINE-->

smith committed:
Add todos

smith committed:
more

smith committed:
Merge remote-tracking branch 'upstream/master' into nls/lang-trace

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/70570/)
* Commit: e246275fb575750cb26b1dfbc0ed25b346d93f66
* This comment will update when the build is complete

### Failed CI Steps
* [Execute oss-pluginFunctional](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/70570/execution/node/471/log)
* [Execute oss-pluginFunctional](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/70570/execution/node/582/log)

### History
* :broken_heart: [Build #70381](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/70381/) failed 21ec6b04a5040d35c5637e4c7420c3bebc604bbc

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/70381/","number":"70381","commit":"21ec6b04a5040d35c5637e4c7420c3bebc604bbc"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/70570/","number":"70570","commit":"e246275fb575750cb26b1dfbc0ed25b346d93f66"}]}
PIPELINE-->

smith committed:
add at prefix for java

elasticmachine commented:
Pinging @elastic/apm-ui (Team:apm)

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/70596/)
* Commit: 2741509b015389a00ca60ef5a054f8467efff363

### Build metrics
#### @kbn/optimizer bundle module count
| id | [value](https://github.com/elastic/kibana/commit/2741509b015389a00ca60ef5a054f8467efff363) | diff | [baseline](https://github.com/elastic/kibana/commit/5a9d227eee1b53673c6445c00746a0846bb69e48) |
| --- | --- | --- | --- |
| apm | 949 | +6 | 943 |

#### async chunks size
| id | [value](https://github.com/elastic/kibana/commit/2741509b015389a00ca60ef5a054f8467efff363) | diff | [baseline](https://github.com/elastic/kibana/commit/5a9d227eee1b53673c6445c00746a0846bb69e48) |
| --- | --- | --- | --- |
| apm | 4.7MB | +9.0KB | 4.7MB |

#### distributable file count
| id | [value](https://github.com/elastic/kibana/commit/2741509b015389a00ca60ef5a054f8467efff363) | diff | [baseline](https://github.com/elastic/kibana/commit/5a9d227eee1b53673c6445c00746a0846bb69e48) |
| --- | --- | --- | --- |
| total | 53203 | +6 | 53197 |

### History
* :broken_heart: [Build #70570](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/70570/) failed e246275fb575750cb26b1dfbc0ed25b346d93f66
* :broken_heart: [Build #70381](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/70381/) failed 21ec6b04a5040d35c5637e4c7420c3bebc604bbc

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/70381/","number":"70381","commit":"21ec6b04a5040d35c5637e4c7420c3bebc604bbc"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/70570/","number":"70570","commit":"e246275fb575750cb26b1dfbc0ed25b346d93f66"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/70596/","number":"70596","commit":"2741509b015389a00ca60ef5a054f8467efff363"}]}
PIPELINE-->

smith committed:
[APM] Fix overlapping transaction names

...in the table and the header. Did this by adding `word-break: break-all` to them.

Also:

* Rename List to TransactionList
* Add stories for TransactionList and ApmHeader
* Add missing type information to transactions based on sample transaction
*
Fixes #73960.

smith committed:
Revert "[APM] Fix overlapping transaction names"

This reverts commit 1ef000e0fc728161372ef1ef45eded5e961375ca.

smith committed:
Merge remote-tracking branch 'upstream/master' into nls/lang-trace

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/70848/)
* Commit: 7e8d01e2c8932702049952e85628d31800079763

### Build metrics
#### @kbn/optimizer bundle module count
| id | [value](https://github.com/elastic/kibana/commit/7e8d01e2c8932702049952e85628d31800079763) | diff | [baseline](https://github.com/elastic/kibana/commit/d457d530017e086064ed174fc74f9a8d738bd6bb) |
| --- | --- | --- | --- |
| apm | 949 | +6 | 943 |

#### async chunks size
| id | [value](https://github.com/elastic/kibana/commit/7e8d01e2c8932702049952e85628d31800079763) | diff | [baseline](https://github.com/elastic/kibana/commit/d457d530017e086064ed174fc74f9a8d738bd6bb) |
| --- | --- | --- | --- |
| apm | 4.7MB | +9.0KB | 4.7MB |

#### distributable file count
| id | [value](https://github.com/elastic/kibana/commit/7e8d01e2c8932702049952e85628d31800079763) | diff | [baseline](https://github.com/elastic/kibana/commit/d457d530017e086064ed174fc74f9a8d738bd6bb) |
| --- | --- | --- | --- |
| total | 53132 | +6 | 53126 |

### History
* :green_heart: [Build #70596](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/70596/) succeeded 2741509b015389a00ca60ef5a054f8467efff363
* :broken_heart: [Build #70570](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/70570/) failed e246275fb575750cb26b1dfbc0ed25b346d93f66
* :broken_heart: [Build #70381](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/70381/) failed 21ec6b04a5040d35c5637e4c7420c3bebc604bbc

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/70381/","number":"70381","commit":"21ec6b04a5040d35c5637e4c7420c3bebc604bbc"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/70570/","number":"70570","commit":"e246275fb575750cb26b1dfbc0ed25b346d93f66"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/70596/","number":"70596","commit":"2741509b015389a00ca60ef5a054f8467efff363"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/70848/","number":"70848","commit":"7e8d01e2c8932702049952e85628d31800079763"}]}
PIPELINE-->

smith committed:
Merge remote-tracking branch 'upstream/master' into nls/lang-trace

smith committed:
Formatting

smith commented:
@felixbarny wrote (on internal Elastic Slack):

> Some more feedback: before the exception message there should also be the exception type. 

I did not change this. We're using the `message` here. I think if the type is part of the message that's what we should store. So far in this PR we're only adding language-specific formatting for the stackframe headings, not for the messages.

I agree with you it should show the class in the message, but I'm not sure this is the right place to do it.

> Could we decrease the padding in-between the frames?

Done.

@formgeist this includes some changes to the spacing and indentation. I'm adding you as a reviewer.

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/71041/)
* Commit: 21b9f2ac94d638623f25b00635f131b6530dcee5

### Build metrics
#### @kbn/optimizer bundle module count
| id | [value](https://github.com/elastic/kibana/commit/21b9f2ac94d638623f25b00635f131b6530dcee5) | diff | [baseline](https://github.com/elastic/kibana/commit/d556c79481e7fa19b3644f8e925ea0236e237a44) |
| --- | --- | --- | --- |
| apm | 949 | +6 | 943 |

#### async chunks size
| id | [value](https://github.com/elastic/kibana/commit/21b9f2ac94d638623f25b00635f131b6530dcee5) | diff | [baseline](https://github.com/elastic/kibana/commit/d556c79481e7fa19b3644f8e925ea0236e237a44) |
| --- | --- | --- | --- |
| apm | 4.7MB | +8.9KB | 4.7MB |

#### distributable file count
| id | [value](https://github.com/elastic/kibana/commit/21b9f2ac94d638623f25b00635f131b6530dcee5) | diff | [baseline](https://github.com/elastic/kibana/commit/d556c79481e7fa19b3644f8e925ea0236e237a44) |
| --- | --- | --- | --- |
| total | 53133 | +7 | 53126 |

### History
* :green_heart: [Build #70848](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/70848/) succeeded 7e8d01e2c8932702049952e85628d31800079763
* :green_heart: [Build #70596](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/70596/) succeeded 2741509b015389a00ca60ef5a054f8467efff363
* :broken_heart: [Build #70570](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/70570/) failed e246275fb575750cb26b1dfbc0ed25b346d93f66
* :broken_heart: [Build #70381](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/70381/) failed 21ec6b04a5040d35c5637e4c7420c3bebc604bbc

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/70381/","number":"70381","commit":"21ec6b04a5040d35c5637e4c7420c3bebc604bbc"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/70570/","number":"70570","commit":"e246275fb575750cb26b1dfbc0ed25b346d93f66"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/70596/","number":"70596","commit":"2741509b015389a00ca60ef5a054f8467efff363"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/70848/","number":"70848","commit":"7e8d01e2c8932702049952e85628d31800079763"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/71041/","number":"71041","commit":"21b9f2ac94d638623f25b00635f131b6530dcee5"}]}
PIPELINE-->

sorenlouv reviewed (approved):
The spacing and indentation make it look so much better. And the language-specific changes give the stack traces a native feel. Great improvement!

smith commented:
@elasticmachine merge upstream

felixbarny commented:
> I did not change this. We're using the message here. I think if the type is part of the message that's what we should store. So far in this PR we're only adding language-specific formatting for the stackframe headings, not for the messages.

I think we should store the exception type and message in different fields but then display both.

> I agree with you it should show the class in the message, but I'm not sure this is the right place to do it.

You mean this PR is not the right place to do it? I'm fine either way but when not adding it in this PR, could you create a follow-up issue? But if it's a quick thing to do I'm not sure if the overhead of another issue is worth it. Up to you 🙂

smith commented:
> You mean this PR is not the right place to do it? I'm fine either way but when not adding it in this PR, could you create a follow-up issue? But if it's a quick thing to do I'm not sure if the overhead of another issue is worth it. Up to you 🙂

Yes I'd like to do it in another PR, since this is about language-specific stackframes and this is making changes to messages. I'll open an issue.

smith commented:
@elasticmachine merge upstream

elasticmachine committed:
Merge branch 'master' into nls/lang-trace

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/71266/)
* Commit: f994810f72523a05c00bf3e8f0e26808dcc5fe4a

### Build metrics
#### @kbn/optimizer bundle module count
| id | [value](https://github.com/elastic/kibana/commit/f994810f72523a05c00bf3e8f0e26808dcc5fe4a) | diff | [baseline](https://github.com/elastic/kibana/commit/9b5912203f2211bd16f871a5f05cb59f1cae173a) |
| --- | --- | --- | --- |
| apm | 947 | +6 | 941 |

#### async chunks size
| id | [value](https://github.com/elastic/kibana/commit/f994810f72523a05c00bf3e8f0e26808dcc5fe4a) | diff | [baseline](https://github.com/elastic/kibana/commit/9b5912203f2211bd16f871a5f05cb59f1cae173a) |
| --- | --- | --- | --- |
| apm | 4.7MB | +8.9KB | 4.7MB |

### History
* :green_heart: [Build #71041](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/71041/) succeeded 21b9f2ac94d638623f25b00635f131b6530dcee5
* :green_heart: [Build #70848](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/70848/) succeeded 7e8d01e2c8932702049952e85628d31800079763
* :green_heart: [Build #70596](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/70596/) succeeded 2741509b015389a00ca60ef5a054f8467efff363
* :broken_heart: [Build #70570](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/70570/) failed e246275fb575750cb26b1dfbc0ed25b346d93f66
* :broken_heart: [Build #70381](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/70381/) failed 21ec6b04a5040d35c5637e4c7420c3bebc604bbc

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/70381/","number":"70381","commit":"21ec6b04a5040d35c5637e4c7420c3bebc604bbc"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/70570/","number":"70570","commit":"e246275fb575750cb26b1dfbc0ed25b346d93f66"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/70596/","number":"70596","commit":"2741509b015389a00ca60ef5a054f8467efff363"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/70848/","number":"70848","commit":"7e8d01e2c8932702049952e85628d31800079763"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/71041/","number":"71041","commit":"21b9f2ac94d638623f25b00635f131b6530dcee5"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/71266/","number":"71266","commit":"f994810f72523a05c00bf3e8f0e26808dcc5fe4a"}]}
PIPELINE-->

cauemarcondes commented:
New bug: https://github.com/elastic/kibana/issues/80152
Other than that it works great.
Chrome ✅
FF ✅
Safari ✅
